benches/dual_numbers.rs

```diff
@@ -0,0 +1,233 @@
+//! Benchmarks for the evaluation of the Helmholtz energy function
+//! for a given `StateHD` for different types of dual numbers.
+//! These should give an idea about the expected slow-down depending
+//! on the dual number type used without the overhead of the `State`
+//! creation.
+use criterion::{criterion_group, criterion_main, Criterion};
+use feos::pcsaft::{PcSaft, PcSaftParameters};
+use feos_core::{
+    parameter::{IdentifierOption, Parameter},
+    EquationOfState, HelmholtzEnergy, HelmholtzEnergyDual, State, StateHD,
+};
+use ndarray::arr1;
+use num_dual::{Dual3, Dual64, DualNum, HyperDual64};
+use quantity::si::*;
+use std::sync::Arc;
+
+/// Residual Helmholtz energy given a StateHD
+fn a_res<D: DualNum<f64>>(inp: (&Arc<PcSaft>, &StateHD<D>)) -> D
+where
+    (dyn HelmholtzEnergy + 'static): HelmholtzEnergyDual<D>,
+{
+    inp.0.evaluate_residual(inp.1)
+}
+
+fn methane_pcsaft(c: &mut Criterion) {
+    let parameters = PcSaftParameters::from_json(
+        vec!["methane"],
+        "./parameters/pcsaft/gross2001.json",
+        None,
+        IdentifierOption::Name,
+    )
+    .unwrap();
+    let eos = Arc::new(PcSaft::new(Arc::new(parameters)));
+
+    let cp = State::critical_point(&eos, None, None, Default::default()).unwrap();
+
+    let t = 0.8 * cp.temperature;
+    let density = cp.density;
+    let volume = 10.0 * MOL / density;
+    let x = arr1(&[1.0]);
+    let m = &x * 10.0 * MOL;
+
+    let mut group = c.benchmark_group("PC-SAFT methane");
+
+    // real valued evaluation
```

Comment:
It looks like the calculation of the four cases could be reused between the different case studies